libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp

```diff
@@ -0,0 +1,469 @@
+    // shared state guards access to the next stage; when the shared state goes
+    // out of scope it triggers continuations for the next stage.
+    //
+    // When read-write access is required a sender is created which holds on to
+    // the newly created shared state for the read-write access and the previous
+    // state. When the sender is connected to a receiver, a callback is added to
+    // the previous shared state's destructor. The callback holds the new state,
+    // and passes a wrapper holding the shared state to set_value. Once the
+    // receiver which receives the wrapper has let the wrapper go out of scope
+    // (and all other references to the shared state are out of scope), the new
+    // shared state will again trigger its continuations.
+    //
+    // When read-only access is required and the previous access was read-only
+    // the procedure is the same as for read-write access. When read-only access
+    // follows a previous read-only access the shared state is reused between
+    // all consecutive read-only accesses, such that multiple read-only accesses
+    // can run concurrently, and the next access (which must be read-write) is
+    // triggered once all instances of that shared state have gone out of scope.
+    //
+    // The protected value is heap-allocated and reference counted (in a
+    // shared_ptr) to allow it to live past the destructor of the mutex, and to
+    // avoid moving the value between each stage.
+
+    template <typename Allocator>
+    class async_rw_mutex<void, void, Allocator>
+    {
+    private:
+        template <detail::async_rw_mutex_access_type AccessType>
+        struct sender;
+
+    public:
+        using read_type = void;
+        using readwrite_type = void;
+
+        using read_access_type =
+            detail::async_rw_mutex_access_wrapper<readwrite_type, read_type,
+                detail::async_rw_mutex_access_type::read>;
+        using readwrite_access_type =
+            detail::async_rw_mutex_access_wrapper<readwrite_type, read_type,
+                detail::async_rw_mutex_access_type::readwrite>;
+
+        using allocator_type = Allocator;
+
+        explicit async_rw_mutex(Allocator const& alloc = {})
+          : alloc(alloc){};
+        async_rw_mutex(async_rw_mutex&&) = default;
+        async_rw_mutex& operator=(async_rw_mutex&&) = default;
+        async_rw_mutex(async_rw_mutex const&) = delete;
+        async_rw_mutex& operator=(async_rw_mutex const&) = delete;
+
+        sender<detail::async_rw_mutex_access_type::read> read()
+        {
+            HPX_ASSERT(state);
+
+            if (prev_access == detail::async_rw_mutex_access_type::readwrite)
+            {
+                prev_state = state;
+                state =
+                    std::allocate_shared<detail::async_rw_mutex_shared_state,
+                        Allocator>(alloc);
+                prev_access = detail::async_rw_mutex_access_type::read;
```

Comment:
Can the disjoint testing and setting of `prev_access` lead to races?

Reply:
Yes and no, meaning it's intentionally not threadsafe. It's mentioned in the docstring. Not sure if it's worth making it threadsafe.